accepted/wasmtime-debugging-v2.md

```diff
@@ -0,0 +1,533 @@
+- record the fiber state (saved SP).
+  - (note that the fiber library itself saves all callee-saved
+    registers to the stack before switching away, so we don't need to
+    handle those separately)
+
+And the restore procedure is the opposite:
+
+- `memcpy` from the snapshot of the stack back onto the stack;
+- resume to the fiber.
+
+Notably, and perhaps surprisingly, all of this can actually be
+implemented purely in the fiber library: so we have a general notion
+of fiber snapshots (almost like setjmp/longjmp primitives on fibers).
+
+## Record/Replay, Store Boundary, and Host API
+
+Arjun Ramesh (@arjunr2) has implemented a [prototype of record/replay
+in Wasmtime](https://github.com/bytecodealliance/wasmtime/pull/11284),
+proving the viability of this component of our overall time-travel
+debugging plans with remarkably low overhead (~4% execution time
+increase). For completeness, we describe a few high-level aspects of
+the design and interface here, and note its relation to the above
+plans.
+
+The general simplifying assumption behind the design is that we record
+*an entire store*, and trace all interactions across the store
+boundary so that we can replay them. The key choice that makes this
+tracing feasible is that we focus on recording *Wasm components*,
+which by specification must not export their memory; shared-memory
+interactions are the bane of any tracing scheme, and this approach
+manages the complexity by instead hooking the explicit
+canonical-ABI-specified writes of structured data into the component's
+linear memory.
+
+The prototype linked above provides trace and replay modes via
+Wasmtime's `Config`, but requires the host to initiate calls and other
+interactions with the store; calls from the component back *out* to
+the host are recorded and replayed. In the final form, we plan to
+record *both* directions, i.e., the host's calls into Wasm (recording
+the arguments to calls), and the Wasm's calls back to the host
+(recording the return values from calls). We will then provide two new
+abstractions at the host API level: a `Trace`, which is analogous to a
+`Module` as a static entity that can be executed, and a `Replay`,
+which is analogous to an `Instance` as one dynamic execution of a
+`Trace`. The main difference is that the `Replay` will fully own its
+`Store`, rather than simply being instantiated with an existing store,
+because it must fully control all actions within the store to ensure
+determinism.
+
+Once we have this infrastructure, we can provide debugging APIs on top
+of it the same way we will for live instances. Note that the two
+implementation efforts can proceed largely independently until we
+combine them. In particular, also, note that *the recording need not
+run the same machine code as the replay*: this is critical, because it
+means that we can keep the same fast recording overhead in production,
+and pay the cost of debugging instrumentation only when examining a
+trace of a bug.
```

Comment:
💯 

Fast (optimized) recording without requiring full (deoptimized, slow) state preservation until replay time is a must.